File: week11/7/[241113](Leet1584)_Min Cost to Connect All Points.py
"""You are given an array points representing integer coordinates of some points on a 2D-plane, where points[i] = [xi, yi].

The cost of connecting two points [xi, yi] and [xj, yj] is the manhattan distance between them: |xi - xj| + |yi - yj|, where |val| denotes the absolute value of val.

Return the minimum cost to make all points connected. All points are connected if there is exactly one simple path between any two points"""

"""Minimum Spanning Tree 문제 : Kruskal 또는 prim으로 풀 수 있다. 나는 Kruskal로 품."""


# Kruskal 풀이
class DisjointSet:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, u):
        if self.parent[u] != u:
            self.parent[u] = self.find(self.parent[u])  # Path compression
        return self.parent[u]

    def union(self, u, v):
        root_u = self.find(u)
        root_v = self.find(v)
        if root_u != root_v:
            # Union by rank
            if self.rank[root_u] > self.rank[root_v]:
                self.parent[root_v] = root_u
            elif self.rank[root_u] < self.rank[root_v]:
                self.parent[root_u] = root_v
            else:
                self.parent[root_v] = root_u
                self.rank[root_u] += 1


class Solution(object):
    def kruskal(self, n, edges):
        dsu = DisjointSet(n)
        mst = []
        edges.sort(key=lambda x: x[2])  # Sort edges by weight
        
        for u, v, weight in edges:
            if dsu.find(u) != dsu.find(v):
                dsu.union(u, v)
                mst.append((u, v, weight))
        
        return mst

    #//절댓값(맨해튼 거리) 계산용으로 만든 함수
    def make_edge_point_index(self, i, j, points) :
        p1 = points[i]
        p2 = points[j]
        
        w_x = p1[0]-p2[0]
        w_y = p1[1]-p2[1]
        w_x = -w_x if w_x < 0 else w_x
        w_y = -w_y if w_y < 0 else w_y
        
        w = w_x + w_y
        
        return (i, j, w)
    
    def minCostConnectPoints(self, points):
        edges = []
        points_len = len(points)
        for i in range(points_len):
            for j in range(points_len):
                if i != j:
                    edges.append(self.make_edge_point_index(i, j, points))
        
        mst_edges = self.kruskal(points_len, edges)
        
        sum = 0
        for i in range(len(mst_edges)):
            sum += mst_edges[i][2]
        
        return sum
    
    
# Prim 풀이 참고.
import heapq
class Solution:
    def minCostConnectPoints(self, points):
        n = len(points)
        # for simplicity, marking the points with a node value from 0 to (n-1)
        adjList = {i:[] for i in range(n)}  # key = curr node, value = (dist, other nodes)
        for i in range(n):
            x1 = points[i][0]; y1 = points[i][1]
            for j in range(i+1, n):
                x2 = points[j][0]; y2 = points[j][1]
                dist = abs(x1 - x2) + abs(y1 - y2)
                # as it is a undirected graph, we need to add both i and j to adjList
                adjList[i].append((dist, j))
                adjList[j].append((dist, i))
        
        # Prim's Algorithm - Minimum Spanning Tree (MST) 
        # (In the Dijkstra's Algorithm if we introduce a visited data structure to 
        # stop cycle in graph then it converts to Prim's Algorithm)
        visited = set()
        res = 0
        minHeap = [(0, 0)]  # (dist, node)
        while len(visited) < n: # until all nodes a visited
            dist, node = heapq.heappop(minHeap)
            if node in visited: continue
            visited.add(node)
            res += dist
            for dn in adjList[node]:  # dn = (dist, node)
                if dn[1] not in visited:
                    heapq.heappush(minHeap, dn)
        
        return res


# V = number of umber of verices and E = numver of Edges in the graph.
# Time: O(V^2) + O((V+E)logV) = O(V^2)
# Space: O(V^2)  In worst cases all connections of undirected graph can be in minheap 

File: week11/7/[241113](Leet1584)_Min Cost to Connect All Points.cpp
/*You are given an array points representing integer coordinates of some points on a 2D-plane, where points[i] = [xi, yi].

The cost of connecting two points [xi, yi] and [xj, yj] is the manhattan distance between them: |xi - xj| + |yi - yj|, where |val| denotes the absolute value of val.

Return the minimum cost to make all points connected. All points are connected if there is exactly one simple path between any two points*/

#include <iostream>
#include <vector>
#include <algorithm>

struct Edge {
    int u, v, weight;
    bool operator<(const Edge& other) const {
        return weight < other.weight;
    }
};

struct DisjointSet {
    std::vector<int> parent, rank;
    DisjointSet(int n) : parent(n), rank(n, 0) {
        for (int i = 0; i < n; ++i) parent[i] = i;
    }

    int find(int u) {
        if (u != parent[u]) parent[u] = find(parent[u]);  // Path compression
        return parent[u];
    }

    void unionSet(int u, int v) {
        int root_u = find(u);
        int root_v = find(v);
        if (root_u != root_v) {
            if (rank[root_u] > rank[root_v]) parent[root_v] = root_u;
            else if (rank[root_u] < rank[root_v]) parent[root_u] = root_v;
            else {
                parent[root_v] = root_u;
                rank[root_u]++;
            }
        }
    }
};

std::vector<Edge> kruskal(int n, std::vector<Edge>& edges) {
    std::sort(edges.begin(), edges.end());
    DisjointSet dsu(n);
    std::vector<Edge> mst;
    
    for (auto& edge : edges) {
        if (dsu.find(edge.u) != dsu.find(edge.v)) {
            dsu.unionSet(edge.u, edge.v);
            mst.push_back(edge);
        }
    }
    
    return mst;
}

//절댓값(맨해튼 거리) 계산용으로 만든 함수
Edge make_edge_with_index(int i, int j, const std::vector<std::vector<int>>& points){
    int w_x = (points[i][0] - points[j][0]) > 0 ? points[i][0]-points[j][0] : - (points[i][0] - points[j][0]);
    int w_y = (points[i][1] - points[j][1]) > 0 ? points[i][1]-points[j][1] : - (points[i][1] - points[j][1]);

    Edge new_edge;
    new_edge.u = i;
    new_edge.v = j;
    new_edge.weight = w_x + w_y;

    return new_edge;
}

class Solution {
public:
    int minCostConnectPoints(std::vector<std::vector<int>>& points) {
        std::vector<Edge> edge_list;
        
        int num_points = points.size();
        for (int i = 0; i < num_points; i++){
            for (int j = 0; j < num_points; j++){
                edge_list.push_back(make_edge_with_index(i, j, points));
            }
        }

        std::vector<Edge> MST_edge = kruskal(num_points, edge_list);
        
        int sum = 0;
        for (auto e : MST_edge){
            sum += e.weight;
        }
        return sum;
    }
};


//Prim :참고
#include <vector>
#include <queue>
#include <utility>
#include <cmath>     // 혹은 #include <cstdlib> (절대값 함수 abs 사용)
#include <functional>  // 비교함수 greater<>를 사용할 때 필요
using namespace std;

class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& nums) {
        int n=nums.size();
        int connected = 0, i=0;
        int ans=0;
        vector<bool> vis(n);
        priority_queue<pair<int,int>, vector<pair<int,int>>,
        greater<pair<int,int>>> pq;
        while(++connected < n){
            vis[i] = true;
            for(int j=0;j<n;j++){
                if(!vis[j]) 
                    pq.push({(abs(nums[j][0] - nums[i][0]) + 
                    abs(nums[i][1] - nums[j][1])), j});
            }

            while(vis[pq.top().second]) pq.pop();
            ans=ans+=pq.top().first;
            i = pq.top().second;
            pq.pop();
        }
        return ans;
    }
};